### PR TITLE
Fixed #1015 - Teams user registration is broken

### LINKED ISSUE #1015: Registration upon user invite through Teams is broken
### Issue Summary

When someone is invited to join a team, the join link redirects to settings instead of sign up, as a new user of the team am not able to signup.

### Steps to Reproduce

1. Invite someone to a team.
2. On getting redirected to sign up page, fill in form.
3. Press "Sign up"
3. It says user already registered.

image

### PR DESCRIPTION
**Pull request by emrysal**
The necessary check got removed in https://github.com/calendso/calendso/commit/13486d9988a11a93eeee85b0396a57c0dbad43d4#diff-bc80461208be7df84b92ad01439aea7d4dcee60b443e1ba998a77cebed65c1e2

Continued expanding the `fields.tsx` file with additional features; deprecating UsernameInput in the process. I tried to keep the implementation side as lean as possible.

As soon as /availability is merged we can replace `<form>` with `<Form>` and we'll no longer need to manually import FormProvider, neatening it up further.

Note: This refactor focusses on react-hook-forms, it does not remove the fetch (one step at a time..)

### CODE REVIEW

**Review by pumfleet — APPROVED** [review]
Looks good to me. Hopefully we can get the /availability revamp merged today, so you can update this PR to use <Form> and then merge

**Comment by baileypumfleet** [comment]
Now that the /availability PR is merged, the modification for this PR can be made

File: components/form/fields.tsx [hunk 1]
```diff
@@ -28,78 +31,97 @@ export function Label(props: JSX.IntrinsicElements["label"]) {
...
         {props.children}
       </form>
     </FormProvider>
   );
-}
+};
+
+export const Form = forwardRef(PlainForm) as <T extends FieldValues>(
+  p: FormProps<T> & { ref?: Ref<HTMLFormElement> }
+) => ReactElement;
```
Comment (zomars) [inline components/form/fields.tsx:124]: Preserve form value types from generics

File: pages/getting-started.tsx [hunk 2]
```diff
@@ -313,7 +313,7 @@ export default function Onboarding(props: inferSSRProps<typeof getServerSideProp
       title: t("set_availability"),
       description: t("set_availability_instructions"),
       Component: (
-        <Form
+        <Form<ScheduleFormValues>
```
Comment (zomars) [inline pages/getting-started.tsx:316]: So our values in `handleSubmit` are typed.

**Review by zomars — APPROVED** [review]
Made some fixes to the onboarding. I think this is ready to merge.

File: pages/api/schedule/index.ts [hunk 3]
```diff
@@ -48,27 +48,29 @@ export default async function handler(req: NextApiRequest, res: NextApiResponse)
...
+            createMany: {
+              data: availability.map((schedule) => ({
+                days: schedule.days,
+                startTime: schedule.startTime,
+                endTime: schedule.endTime,
+              })),
+            },
+          },
+        },
+      });
```
Comment (zomars) [inline pages/api/schedule/index.ts:73]: Made this query safer in more legible (IMO) Tested and working 👍🏽

**Review by zomars — APPROVED** [review]

**Comment by emrysal** [comment]
Feel free to merge when ready @zomars - most of the content in this PR has been discussed at length and I'll tweak using the fields.tsx modifications later.

File: components/ui/form/Schedule.tsx [hunk 4]
```diff
@@ -30,22 +31,6 @@ const TIMES = (() => {
...
-export const DEFAULT_SCHEDULE: ScheduleType = [
-  [],
-  [defaultDayRange],
-  [defaultDayRange],
-  [defaultDayRange],
-  [defaultDayRange],
-  [defaultDayRange],
-  [],
-];
-
```
Comment (zomars) [inline components/ui/form/Schedule.tsx:48]: Moved so we can reuse in the seeder

File: pages/api/schedule/index.ts [hunk 5]
```diff
@@ -17,58 +16,33 @@ export default async function handler(req: NextApiRequest, res: NextApiResponse)
...
-        }
-        return true;
-      });
-      filteredTimes.forEach((time) => {
-        availability.push(addNewTime(time));
-      });
-      return availability;
-    },
-    [] as Availability[]
-  );
```
Comment (zomars) [inline pages/api/schedule/index.ts:47]: Moved so we can reuse the logic in the seeder

File: scripts/seed.ts [hunk 6]
```diff
@@ -26,6 +27,11 @@ async function createUserAndEventType(opts: {
     password: await hashPassword(opts.user.password),
     emailVerified: new Date(),
     completedOnboarding: opts.user.completedOnboarding ?? true,
+    availability: {
+      createMany: {
+        data: getAvailabilityFromSchedule(DEFAULT_SCHEDULE),
+      },
+    },
```
Comment (zomars) [inline scripts/seed.ts:34]: Tests were failing because there were no availability in a fresh DB

**Review by zomars — APPROVED** [review]

**Comment by zomars** [comment]
Merged and going to sleep 👍🏽

**Pull request merged**